systemtest/benchtest/expvarmetrics/metrics.go

```diff
@@ -0,0 +1,253 @@
+// event that was collected by the collector.
+func (c *Collector) Delta(m Metric) int64 {
+	c.l.Lock()
+	defer c.l.Unlock()
+
+	stats := c.metrics[m]
+	return stats.Last - stats.First
+}
+
+// AddWatch configures a new watcher for a given metric. The
+// watcher is deleted after the specified expected value is
+// observed.
+//
+// Watcher returns a read only channel to observe the state
+// of the watch. A true event on this channel refers to an
+// observation with the expected value while a false event
+// refers to an error in the collector. Both cases end the
+// lifecycle of a watcher.
+func (c *Collector) AddWatch(m Metric, expected int64) (<-chan bool, error) {
+	c.l.Lock()
+	defer c.l.Unlock()
+
+	if _, ok := c.watchers[m]; ok {
+		return nil, errors.New("watcher already exists for the given metric")
+	}
+
+	out := make(chan bool)
+	c.watchers[m] = &watchItem{
+		notifyChan: out,
+		validator:  func(in int64) bool { return in == expected },
+	}
+
+	return out, nil
+}
+
+func (c *Collector) addExpvar(e expvar) {
+	c.l.Lock()
+	defer c.l.Unlock()
+
+	c.processMetric(Goroutines, e.Goroutines)
+	c.processMetric(Bytes, e.UncompressedBytes)
+	c.processMetric(TotalEvents, e.TotalEvents)
+	c.processMetric(ActiveEvents, e.ActiveEvents)
+	c.processMetric(RSSMemoryBytes, e.RSSMemoryBytes)
+	c.processMetric(AvailableBulkRequests, e.AvailableBulkRequests)
+	c.processMetric(TransactionsProcessed, e.TransactionsProcessed)
+	c.processMetric(SpansProcessed, e.SpansProcessed)
+	c.processMetric(MetricsProcessed, e.MetricsProcessed)
+	c.processMetric(ErrorsProcessed, e.ErrorsProcessed)
+	c.processMetric(ErrorElasticResponses, e.ErrorElasticResponses)
+	c.processMetric(ErrorOTLPTracesResponses, e.ErrorOTLPTracesResponses)
+	c.processMetric(ErrorOTLPMetricsResponses, e.ErrorOTLPMetricsResponses)
+	c.processMetric(NumGC, int64(e.NumGC))
+	c.processMetric(PauseTotalNs, int64(e.PauseTotalNs))
+	c.processMetric(MemAllocs, int64(e.Mallocs))
+	c.processMetric(MemBytes, int64(e.TotalAlloc))
+	c.processMetric(HeapAlloc, int64(e.HeapAlloc))
+	c.processMetric(HeapObjects, int64(e.HeapObjects))
+}
+
+func (c *Collector) processMetric(m Metric, val int64) {
+	if _, ok := c.metrics[m]; !ok {
+		c.metrics[m] = &AggregateStats{}
+	}
+	c.updateMetric(c.metrics[m], val)
+
+	if watcher, ok := c.watchers[m]; ok && watcher.validator(val) {
+		watcher.notifyChan <- true
+		delete(c.watchers, m)
+	}
+}
+
+func (c *Collector) updateMetric(stats *AggregateStats, value int64) {
+	// Initialize for first value
+	if stats.sampleCount == 0 {
+		stats.First = value
+		stats.Min = value
+		stats.Max = value
+	}
+
+	stats.Last = value
+	stats.sampleCount += 1
+	stats.Mean += (float64(value) - stats.Mean) / float64(stats.sampleCount)
+	if stats.Min > value {
+		stats.Min = value
+	}
+	if stats.Max < value {
+		stats.Max = value
+	}
+}
+
+func (c *Collector) rejectWatchers() {
+	c.l.Lock()
+	defer c.l.Unlock()
+
+	for _, watcher := range c.watchers {
+		watcher.notifyChan <- false
+	}
+}
```

Comment:
Since this function is only expected to be called once (from what I can read in the current logic) and its purpose is signal to the channel receiver that the watch isn't going to be updated anymore, I think it would be better to close all the watches's channels to signal this. Otherwise, clients that might be receiving the channel in a loop will never exit (not that we do that now, but provides some safety).

Then, since we're closing the channel and is an operation which can only be performed once (at the risk of panic), we can add a `closed` boolean to the Collector structure to signal this.

```suggestion
func (c *Collector) closeWatches() {
	c.l.Lock()
	defer c.l.Unlock()

	if c.closed {
		return
	}
	for _, watcher := range c.watches {
		close(watcher.notifyChan)
	}
	c.closed = true
}
```

Reply:
I have updated the PR with code to delete watches, let me know what you think.